File: .github/scripts/ai-review.ts
import { config } from "dotenv";
import { execSync } from "child_process";
import { readFileSync, existsSync } from "fs";
import { join } from "path";
import { Octokit } from "@octokit/rest";
import { OpenAI } from "openai";
import { OpenAI as LlamaOpenAI } from "@llamaindex/openai";

config();

const PROJECT_ROOT = __dirname;
const PROMPT_TEMPLATE_PATH = join(
  PROJECT_ROOT,
  "..",
  "prompts",
  "code_review_prompt.md"
);
const PROMPT_CLASSIFICATION_PATH = join(
  PROJECT_ROOT,
  "..",
  "prompts",
  "classification_prompt.md"
);
const FE_GUIDELINES_PATH = join(PROJECT_ROOT, "..", "copilot-instructions.md");
// const INDEX_PATH = join(PROJECT_ROOT, "..", "indexes");

/**
 * ファイルを読み込む関数
 */
function loadFile(filePath: string): string | null {
  try {
    return readFileSync(filePath, "utf-8");
  } catch (error) {
    console.error(`File not found: ${filePath}`);
    return null;
  }
}

/**
 * ファイルの差分に基づいてコードガイドラインを取得
 */
async function getCodeGuidelines(
  llamaLLM: LlamaOpenAI,
  filename: string,
  filediff: string
): Promise<string> {
  try {
    // FEガイドラインを読み込む
    const feGuidelines = loadFile(FE_GUIDELINES_PATH) || "";

    // ガイドラインのプロンプトを作成
    const guidelinePrompt = `
    あなたは経験豊富なコードレビュアーです。以下のフロントエンド設計ガイドラインと、ファイル '${filename}' の差分を見て、
    このコードがガイドラインに準拠しているか、また改善すべき点がないか日本語で教えてください。
    特にフォルダ構成、命名規則、コンポーネントの責務分離に注目してください。
    回答は500文字以内に簡潔にまとめてください：

    # フロントエンド設計ガイドライン:
    ${feGuidelines}

    # コードの差分:
    ${filediff}
    `;

    // ガイドラインを生成
    const response = await llamaLLM.complete({
      prompt: guidelinePrompt,
    });

    return response.text ?? "";
  } catch (error) {
    console.error(`Error getting guidelines with LlamaOpenAI: ${error}`);
    return "";
  }
}

/**
 * diffテキストをファイル単位に分割する関数
 */
function splitDiffByFile(diffText: string): Record<string, string> {
  const fileDiffs: Record<string, string> = {};
  const lines = diffText.split("\n");
  let currentFile: string | null = null;
  let currentLines: string[] = [];

  for (const line of lines) {
    if (line.startsWith("diff --git a/")) {
      if (currentFile && currentLines.length > 0) {
        fileDiffs[currentFile] = currentLines.join("\n");
      }
      const match = line.match(/^diff --git a\/(.+?) b\/(.+)$/);
      if (match) {
        currentFile = match[1];
        currentLines = [line];
      } else {
        currentFile = null;
        currentLines = [];
      }
    } else {
      if (currentFile !== null) {
        currentLines.push(line);
      }
    }
  }

  if (currentFile && currentLines.length > 0) {
    fileDiffs[currentFile] = currentLines.join("\n");
  }

  return fileDiffs;
}

/**
 * コードレビューを生成する関数
 */
async function generateReview(client: OpenAI, prompt: string): Promise<string> {
  try {
    const response = await client.chat.completions.create({
      model: "gpt-4o",
      messages: [
        { role: "system", content: "You are a great code reviewer." },
        { role: "user", content: prompt },
      ],
      temperature: 0.0,
      max_tokens: 800,
    });
    return (
      response?.choices[0].message.content?.trim() ??
      "Failed to generate review."
    );
  } catch (error) {
    console.error(`Error generating review: ${error}`);
    return "Failed to generate review.";
  }
}

/**
 * レビューのアクション（Approve/Request Changes/Comment）を決定する関数
 */
async function determineAction(
  client: OpenAI,
  reviewContent: string
): Promise<string> {
  try {
    const actionPromptTemplate = loadFile(PROMPT_CLASSIFICATION_PATH);
    if (!actionPromptTemplate) {
      return "Comment";
    }
    const actionPrompt = actionPromptTemplate.replace(
      "{review_content}",
      reviewContent
    );
    const response = await client.chat.completions.create({
      model: "gpt-4o",
      messages: [
        { role: "system", content: "You are a great code reviewer." },
        { role: "user", content: actionPrompt },
      ],
      temperature: 0.0,
      max_tokens: 100,
    });
    return (
      response?.choices[0].message.content?.trim() ??
      "Failed to generate review."
    );
  } catch (error) {
    console.error(`Error determining action: ${error}`);
    return "Comment";
  }
}

/**
 * メイン関数
 */
async function main() {
  const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
  const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
  const repo = process.env.GITHUB_REPOSITORY;
  const eventPath = process.env.GITHUB_EVENT_PATH;

  if (!OPENAI_API_KEY || !GITHUB_TOKEN || !repo || !eventPath) {
    console.error("Required environment variables are not set.");
    return;
  }

  // OpenAIクライアントの初期化
  const client = new OpenAI({ apiKey: OPENAI_API_KEY });

  // LlamaOpenAIの設定
  const llamaLLM = new LlamaOpenAI({
    apiKey: OPENAI_API_KEY,
    temperature: 0.0,
    model: "gpt-4o",
  });

  // Gitの差分を取得
  execSync("git fetch origin main");
  const diffText = execSync("git diff origin/main...HEAD").toString().trim();

  if (!diffText) {
    console.log("No diff to review.");
    return;
  }

  // ファイル単位のdiffマップを作成
  const fileDiffMap = splitDiffByFile(diffText);
  if (Object.keys(fileDiffMap).length === 0) {
    console.log("No file-level diffs found.");
    return;
  }

  // プロンプトテンプレートを読み込む
  const promptTemplate = loadFile(PROMPT_TEMPLATE_PATH);
  if (!promptTemplate) {
    return;
  }

  // ファイルごとのレビュー結果を格納するマップ
  const fileReviewsMap: Record<string, string> = {};

  // ファイルごとにコードレビューを生成
  for (const [filename, filediff] of Object.entries(fileDiffMap)) {
    try {
      // LlamaOpenAIを使用してコードガイドラインを取得
      const guidelines = await getCodeGuidelines(llamaLLM, filename, filediff);

      // プロンプトを生成し、レビューを取得
      const prompt = promptTemplate
        .replace("{diff_text}", filediff)
        .replace("{code_guidelines}", guidelines);

      const fileReview = await generateReview(client, prompt);
      fileReviewsMap[filename] = fileReview;
    } catch (error) {
      console.error(`Error processing file ${filename}: ${error}`);
      fileReviewsMap[filename] = "Failed to generate review.";
    }
  }

  // レビュー内容を整形
  let reviewContent =
    "Nakamura Code Rabbit Review\n# Issues and Fix Suggestions\n";
  for (const [filename, reviewText] of Object.entries(fileReviewsMap)) {
    reviewContent += `\n### ${filename}\n${reviewText}\n`;
  }

  // アクションを決定
  const action = await determineAction(client, reviewContent);
  console.log(`Action: ${action}\nReview Content: ${reviewContent}`);

  // GitHub関連の処理
  const payload = JSON.parse(readFileSync(eventPath, "utf-8"));
  if (!payload.pull_request) {
    console.log("Not a pull_request event.");
    return;
  }

  const prNumber = payload.pull_request.number;
  if (!prNumber) {
    console.log("No PR number found.");
    return;
  }

  // Ocktokitを使用してGitHub APIを呼び出す
  const octokit = new Octokit({ auth: GITHUB_TOKEN });
  const [owner, repoName] = repo.split("/");

  // アクションに基づいてPRにコメントする
  if (action === "Comment") {
    await octokit.issues.createComment({
      owner,
      repo: repoName,
      issue_number: prNumber,
      body: reviewContent,
    });
  } else if (action === "Approve") {
    await octokit.pulls.createReview({
      owner,
      repo: repoName,
      pull_number: prNumber,
      event: "APPROVE",
    });
    await octokit.issues.createComment({
      owner,
      repo: repoName,
      issue_number: prNumber,
      body: reviewContent,
    });
  } else if (action === "Request changes") {
    await octokit.pulls.createReview({
      owner,
      repo: repoName,
      pull_number: prNumber,
      body: reviewContent,
      event: "REQUEST_CHANGES",
    });
  } else {
    await octokit.issues.createComment({
      owner,
      repo: repoName,
      issue_number: prNumber,
      body: reviewContent,
    });
  }
}

// メイン処理の実行
main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: .github/scripts/ai-review.py
import os
import json
import subprocess
from pathlib import Path
import requests
from openai import OpenAI
from llama_index.core import StorageContext, load_index_from_storage
from llama_index.core import Settings
from llama_index.llms.openai import OpenAI as LlamaOpenAI
import re

PROJECT_ROOT = Path(__file__).resolve().parent.parent
PROMPT_TEMPLATE_PATH = PROJECT_ROOT / "prompts/code_review_prompt.md"
PROMPT_CLASSIFICATION_PATH = PROJECT_ROOT / "prompts/classification_prompt.md"
GUIDELINES_PATH = PROJECT_ROOT / "copilot-instructions.md"
INDEX_PATH = PROJECT_ROOT / "indexes"

def load_file(file_path):
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return f.read()
    except FileNotFoundError:
        print(f"File not found: {file_path}")
        return None

def load_index():
    try:
        storage_context = StorageContext.from_defaults(persist_dir=str(INDEX_PATH))
        index = load_index_from_storage(storage_context)
        print("Index loaded successfully.")
        return index
    except Exception as e:
        print(f"Warning: Error loading index: {e}")
        print("Continuing without index...")
        return None

def split_diff_by_file(diff_text):
    file_diffs = {}
    lines = diff_text.splitlines()
    current_file = None
    current_lines = []

    for line in lines:
        if line.startswith("diff --git a/"):
            if current_file and current_lines:
                file_diffs[current_file] = "\n".join(current_lines)
            match = re.match(r"^diff --git a/(.+?) b/(.+)$", line)
            if match:
                filename = match.group(1)
                # TypeScriptファイル（.tsまたは.tsx）のみを対象とする
                if filename.endswith(".ts") or filename.endswith(".tsx"):
                    current_file = filename
                    current_lines = [line]
            else:
                current_file = None
                current_lines = []
        else:
            if current_file is not None:
                current_lines.append(line)

    if current_file and current_lines:
        file_diffs[current_file] = "\n".join(current_lines)

    return file_diffs

def generate_review(client, prompt):
    try:
        review_response = client.chat.completions.create(
            model="gpt-4.1",
            messages=[
                {"role": "system", "content": "You are a great code reviewer."},
                {"role": "user", "content": prompt},
            ],
            temperature=0.0,
            max_tokens=800,
        )
        content = review_response.choices[0].message.content.strip()
        return content
    except Exception as e:
        print(f"Error generating review: {e}")
        return "Failed to generate review."

def determine_action(client, review_content):
    try:
        action_prompt = load_file(PROMPT_CLASSIFICATION_PATH).format(review_content=review_content)
        action_response = client.chat.completions.create(
            model="gpt-4.1",
            messages=[
                {"role": "system", "content": "You are a great code reviewer."},
                {"role": "user", "content": action_prompt},
            ],
            temperature=0.0,
            max_tokens=100,
        )
        action_content = action_response.choices[0].message.content.strip()
        return action_content
    except Exception as e:
        print(f"Error determining action: {e}")
        return "Comment"

def main():
    OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
    REVIEW_GITHUB_TOKEN = os.environ.get("REVIEW_GITHUB_TOKEN")
    repo = os.environ.get("GITHUB_REPOSITORY")
    event_path = os.environ.get("GITHUB_EVENT_PATH")

    if not all([OPENAI_API_KEY, REVIEW_GITHUB_TOKEN, repo, event_path]):
        print("Required environment variables are not set.")
        return

    client = OpenAI(api_key=OPENAI_API_KEY)

    subprocess.run(["git", "fetch", "origin", "main"], check=True)
    diff_result = subprocess.run(
        ["git", "diff", "origin/main...HEAD"], capture_output=True, text=True
    )
    diff_text = diff_result.stdout.strip()

    if not diff_text:
        print("No diff to review.")
        return

    file_diff_map = split_diff_by_file(diff_text)
    if not file_diff_map:
        print("No TypeScript file diffs found. Only .ts and .tsx files are reviewed.")
        return

    prompt_template = load_file(PROMPT_TEMPLATE_PATH)
    if not prompt_template:
        return

    index = load_index()
    # インデックスがない場合
    if not index:
        file_reviews_map = {}
        for filename, filediff in file_diff_map.items():
            try:
                # インデックスがない場合はガイドラインを直接使用
                guidelines = load_file(GUIDELINES_PATH)
                retrieved_guidelines = guidelines if guidelines else "No guidelines available."
                
                prompt = prompt_template.format(
                    diff_text=filediff,
                    code_guidelines=retrieved_guidelines
                )
                file_review = generate_review(client, prompt)
                file_reviews_map[filename] = file_review
            except Exception as e:
                print(f"Error processing file {filename}: {e}")
                file_reviews_map[filename] = "Failed to generate review."
    else:
        # インデックスがある場合
        Settings.llm = LlamaOpenAI(api_key=OPENAI_API_KEY, temperature=0.0)
        Settings.chunk_size = 1024
        query_engine = index.as_query_engine()
        
        file_reviews_map = {}
        for filename, filediff in file_diff_map.items():
            try:
                query = (
                    f"Below is the diff of file '{filename}'. "
                    "Please give me the relevant code guidelines in Japanese:\n"
                    f"{filediff}"
                )
                response = query_engine.query(query)
                retrieved_guidelines = str(response)
                
                prompt = prompt_template.format(
                    diff_text=filediff,
                    code_guidelines=retrieved_guidelines
                )
                file_review = generate_review(client, prompt)
                file_reviews_map[filename] = file_review
            except Exception as e:
                print(f"Error processing file {filename}: {e}")
                file_reviews_map[filename] = "Failed to generate review."

    review_content = "コードレビューを行いました。\n# 修正点と提案\n"
    for filename, review_text in file_reviews_map.items():
        review_content += f"\n### {filename}\n{review_text}\n"

    action = determine_action(client, review_content)
    print(f"Action: {action}\nReview Content: {review_content}")

    with open(event_path, "r", encoding="utf-8") as f:
        payload = json.load(f)

    if "pull_request" not in payload:
        print("Not a pull_request event.")
        return

    pr_number = payload["pull_request"].get("number")
    if not pr_number:
        print("No PR number found.")
        return

    if action == "Comment":
        post_comment_to_pr(repo, pr_number, review_content, REVIEW_GITHUB_TOKEN)
    elif action == "Approve":
        approve_pr(repo, pr_number, REVIEW_GITHUB_TOKEN)
        post_comment_to_pr(repo, pr_number, review_content, REVIEW_GITHUB_TOKEN)
    elif action == "Request changes":
        request_changes_to_pr(repo, pr_number, review_content, REVIEW_GITHUB_TOKEN)
    else:
        post_comment_to_pr(repo, pr_number, review_content, REVIEW_GITHUB_TOKEN)

def post_comment_to_pr(repo, pr_number, body, token):
    url = f"https://api.github.com/repos/{repo}/issues/{pr_number}/comments"
    headers = {
        "Authorization": f"token {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    data = {"body": body}
    resp = requests.post(url, headers=headers, json=data)
    if resp.status_code == 201:
        print("Comment posted successfully.")
    else:
        print(f"Failed to post comment: {resp.status_code} - {resp.text}")

def approve_pr(repo, pr_number, token):
    url = f"https://api.github.com/repos/{repo}/pulls/{pr_number}/reviews"
    headers = {
        "Authorization": f"token {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    data = {"event": "APPROVE"}
    resp = requests.post(url, headers=headers, json=data)
    if resp.status_code == 200:
        print("PR approved successfully.")
    else:
        print(f"Failed to approve PR: {resp.status_code} - {resp.text}")

def request_changes_to_pr(repo, pr_number, body, token):
    url = f"https://api.github.com/repos/{repo}/pulls/{pr_number}/reviews"
    headers = {
        "Authorization": f"token {token}",
        "Accept": "application/vnd.github.v3+json",
    }
    data = {"body": body, "event": "REQUEST_CHANGES"}
    resp = requests.post(url, headers=headers, json=data)
    if resp.status_code == 200:
        print("Request changes successfully.")
    else:
        print(f"Failed to request changes: {resp.status_code} - {resp.text}")

if __name__ == "__main__":
    main()